### ISSUE TITLE
Is it possible to implement CPU operations ?

### ISSUE THREAD

**Issue by SleepProgger** [plaidml/plaidml#378]
I wonder if it is possible to implement Plaidml operations on the CPU (via numpy for example).
This, of course, would be slow, but for a quick test or operations that are hard to implement in Tile it might be "good enough".

I realized i can get the data from an `plaidml.tile.Value` in `Operation.bind` if `var` is set (a `plaidml.Tensor`) when the input is a constant, but if the input is an operation (`source` a `plaidml.tile.Source`) i don't really know how to get the real value out of it.

So to summarize what i want to do is:
At operation executions time get the real value of the input as ndarray, execute a numpy function on it (np.linalg.inv in this specific case) and write it back so the next operation can use it.
Not sure that Is even possible with plaidml ?

**Comment by tzerrell** [comment]
With the current version of PlaidML, my suggestion would be to split the PlaidML computation into two parts at the MatrixInverse rather than writing MatrixInverse as a PlaidML Tile operation. So this would look something like:
```
x = <tensor at end of first half of PlaidML>
x_np = K.get_value(x)
x_inv = np.linalg.inv(x_np)
y = K.variable(x_inv)
<Other PlaidML computations applied to y>
```
I'm not sure if your example would duplicate this behavior or if it would duplicate the first part of the network as you suggested it might.

We've considered making nicer mechanisms for thunking back to the CPU for operations like this as a special (i.e. in Stripe), but we haven't done so yet and we don't have a timeline.

**Comment by SleepProgger** [comment]
Thanks for the answer, but i think your approach would not work for use in custom keras layers (i might be wrong about this).

My above approach *might* be fine. At least if MatrixInverse doesn't declare an input `bind ` is only called once for previous operations.
```
class MatrixInverse(ptile.Operation):
    def __init__(self, x):
        self.x = x
        super(MatrixInverse, self).__init__(None, [], [('O', x.shape)])

    def bind(self, bindings):
        print("Matrix inverse called")
        value = K.get_value(self.x)
        print("Matrix inverse called post get_value")
        tensor = plaidml.Tensor(
            bindings.dev,
            plaidml.Shape(bindings.ctx, self.x.shape.dtype, *self.x.shape.dims))
        with tensor.mmap_discard(bindings.ctx) as view:
            view.copy_from_ndarray(np.linalg.inv(value))
            view.writeback()
        return {'O': tensor}
matrix_inverse = MatrixInverse.function


class Dummy(ptile.Operation):
    def __init__(self, x, name):
        self.x = x
        self._name = name
        code = "function (I) -> (O) { O = I; }"
        super(Dummy, self).__init__(code, [('I', x)], [('O', x.shape)])

    def bind(self, bindings):
        print("Dummy %s called" % self._name)
        return super(Dummy, self).bind(bindings)
dummy = Dummy.function


if __name__ == '__main__':
    data = np.random.normal(size=(4, 2, 2)).astype("float32")
    data_k = K.constant(data, dtype="float32")
    dummy(matrix_inverse(dummy(dummy(data_k, 'inner'), 'inner2')), "outer").eval()
```
=>
```
Matrix inverse called
Dummy inner called
Dummy inner2 called
Matrix inverse called post get_value
Dummy outer called
```

If i declare x as input in MatrixInverse tho (`super(MatrixInverse, self).__init__(None, [('I', x)], [('O', x.shape)])`)
The previous operation seem to be called (or at least bound) in fact twice.

Output with input declared =>
```
Dummy inner called
Dummy inner2 called
Matrix inverse called
Dummy inner called
Dummy inner2 called
Matrix inverse called post get_value
Dummy outer called
```

**Comment by SleepProgger** [comment]
Just in case someone else needs something like this:
```
class CPU_operation(ptile.Operation):
    def __init__(self, inputs, outputs, name=None, side_effects=None):
        self._inputs = inputs
        self._outputs = outputs
        super(CPU_operation, self).__init__(None, [], outputs, name=name, side_effects=side_effects)

    def cpu_operation(self, *args):
        raise NotImplementedError()

    def bind(self, bindings):
        values = [K.get_value(x) for x in self._inputs]
        results = self.cpu_operation(*values)
        ret = dict()
        for data, output in zip(results, self.outputs.items()):
            out_name, out_val = output
            tensor = plaidml.Tensor(bindings.dev, plaidml.Shape(
                bindings.ctx,
                out_val.shape.dtype,
                *out_val.shape.dims)
            )
            with tensor.mmap_discard(bindings.ctx) as view:
                view.copy_from_ndarray(data)
                view.writeback()
            ret[out_name] = tensor
        return ret

# Example
class MatrixInverse(CPU_operation):
    def __init__(self, x, name=None):
        super(MatrixInverse, self).__init__([x], (('O', x.shape),), name=name)

    def cpu_operation(self, x):
        return (np.linalg.inv(x),)
```

It might work, it might burn down your house, i don't know. It seem to work fine though.
A custom TILE implementation should almost always be used but for these cases where that is not an option this might help.
